File: sdk/agentcourt.py
"""
AgentCourt SDK — Submit disputes and get rulings programmatically.
Zero dependencies. Uses standard library only.

Usage:
    from agentcourt import AgentCourt
    
    court = AgentCourt()
    
    ruling = court.dispute(
        claimant="ClientCorp",
        respondent="DevStudio",
        contract={"obligations": ["Build app"], "deadlines": ["2026-07-01T23:59:00Z"]},
        claim="No delivery",
        desired_remedy="Refund",
        policy="freelance-delivery",
        evidence=[
            {"type": "contract", "source": "ClientCorp", "timestamp": "2026-06-01T10:00:00Z",
             "claimed_fact": "No deliverable received", "reliability": "high"}
        ]
    )
    
    print(ruling.confidence)  # HIGH
    print(ruling.remedy)      # full_refund
"""

import json
import urllib.request
import urllib.error
from typing import Optional, List
from dataclasses import dataclass


@dataclass
class Ruling:
    case_id: str
    status: str
    confidence: str
    ruling: str
    reasoning: str
    remedy: str
    facts_established: list
    facts_unknown: list
    matched_rule_id: Optional[str]
    policy_name: Optional[str]
    evidence_scores: list
    ruled_at: str
    
    @property
    def is_high_confidence(self) -> bool:
        return self.confidence == "high"
    
    @property
    def is_ruled(self) -> bool:
        return self.status == "ruled"


class AgentCourt:
    """AgentCourt dispute resolution client. Zero dependencies."""
    
    def __init__(self, base_url: str = "https://agentcourt-api-production.up.railway.app"):
        self.base_url = base_url.rstrip("/")
    
    def _post(self, path: str, data: dict) -> dict:
        payload = json.dumps(data).encode()
        req = urllib.request.Request(
            f"{self.base_url}{path}",
            data=payload,
            method="POST",
            headers={"Content-Type": "application/json"},
        )
        try:
            resp = urllib.request.urlopen(req, timeout=30)
            return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            body = e.read().decode()
            raise Exception(f"API error {e.code}: {body}") from e
    
    def _get(self, path: str) -> dict:
        req = urllib.request.Request(f"{self.base_url}{path}")
        try:
            resp = urllib.request.urlopen(req, timeout=30)
            return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            body = e.read().decode()
            raise Exception(f"API error {e.code}: {body}") from e
    
    def dispute(
        self,
        claimant: str,
        respondent: str,
        contract: dict,
        claim: str,
        desired_remedy: str,
        policy: str = "freelance-delivery",
        evidence: List[dict] = None,
        metadata: dict = None,
    ) -> Ruling:
        """Submit a dispute and get a ruling."""
        payload = {
            "claimant": claimant,
            "respondent": respondent,
            "contract": contract,
            "claim": claim,
            "desired_remedy": desired_remedy,
            "policy": policy,
            "evidence": evidence or [],
            "metadata": metadata or {},
        }
        
        data = self._post("/v1/disputes", payload)
        
        return Ruling(
            case_id=data["case_id"],
            status=data["status"],
            confidence=data["confidence"],
            ruling=data["ruling"],
            reasoning=data.get("reasoning", ""),
            remedy=data["remedy"],
            facts_established=data.get("facts_established", []),
            facts_unknown=data.get("facts_unknown", []),
            matched_rule_id=data.get("matched_rule_id"),
            policy_name=data.get("policy_name"),
            evidence_scores=data.get("evidence_scores", []),
            ruled_at=data.get("ruled_at", ""),
        )
    
    def get_case(self, case_id: str) -> dict:
        """Retrieve a case by ID."""
        return self._get(f"/v1/cases/{case_id}")
    
    def list_policies(self) -> list:
        """List available policy templates."""
        return self._get("/v1/policies")
    
    def health(self) -> dict:
        """Check API health."""
        return self._get("/health")


def dispute(**kwargs) -> Ruling:
    """Quick submit without creating a client."""
    return AgentCourt().dispute(**kwargs)


if __name__ == "__main__":
    print("=== AgentCourt SDK Demo ===\n")
    
    court = AgentCourt()
    
    policies = court.list_policies()
    print(f"Policies: {[p['name'] for p in policies]}")
    
    ruling = court.dispute(
        claimant="Acme Corp",
        respondent="Freelancer X",
        contract={
            "parties": ["Acme Corp", "Freelancer X"],
            "obligations": ["Build dashboard"],
            "deadlines": ["2026-06-15T23:59:00Z"],
            "deliverables": ["React dashboard"],
        },
        claim="Freelancer never delivered the dashboard",
        desired_remedy="Full refund",
        policy="freelance-delivery",
        evidence=[
            {
                "type": "contract",
                "source": "Acme Corp",
                "timestamp": "2026-06-01T10:00:00Z",
                "claimed_fact": "No deliverable received",
                "reliability": "high",
                "content_hash": "sha256:demo123",
            }
        ],
    )
    
    print(f"\nCase: {ruling.case_id}")
    print(f"Status: {ruling.status}")
    print(f"Confidence: {ruling.confidence}")
    print(f"Rule: {ruling.matched_rule_id}")
    print(f"Remedy: {ruling.remedy}")
    print(f"Ruling: {ruling.ruling}")

File: sdk/agentcourt.js
/**
 * AgentCourt JavaScript SDK
 * =========================
 * npm install agentcourt  (coming soon)
 * or copy this file into your project.
 *
 * Usage:
 *   import { AgentCourt } from './agentcourt.js';
 *
 *   const court = new AgentCourt();
 *
 *   const ruling = await court.resolve({
 *     policy: 'freelance-delivery',
 *     claim: 'Work never delivered',
 *     claimant: 'buyer_agent',
 *     respondent: 'freelancer_bot',
 *     desiredRemedy: 'full_refund',
 *     evidence: [...]
 *   });
 *
 *   console.log(ruling.remedy);      // 'full_refund'
 *   console.log(ruling.confidence);  // 'high'
 */

const DEFAULT_API = 'https://agentcourt-api-production.up.railway.app';

export class AgentCourt {
  /**
   * Create an AgentCourt client.
   * @param {string} apiUrl - API base URL
   * @param {number} timeout - Request timeout in ms
   */
  constructor(apiUrl = DEFAULT_API, timeout = 30000) {
    this.apiUrl = apiUrl.replace(/\/$/, '');
    this.timeout = timeout;
  }

  async _post(path, data) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);
    try {
      const res = await fetch(`${this.apiUrl}${path}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(data),
        signal: controller.signal,
      });
      if (!res.ok) {
        const text = await res.text();
        throw new Error(`HTTP ${res.status}: ${text}`);
      }
      return await res.json();
    } finally {
      clearTimeout(timer);
    }
  }

  async _get(path) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);
    try {
      const res = await fetch(`${this.apiUrl}${path}`, {
        signal: controller.signal,
      });
      if (!res.ok) {
        throw new Error(`HTTP ${res.status}`);
      }
      return await res.json();
    } finally {
      clearTimeout(timer);
    }
  }

  /**
   * Submit a dispute and get a ruling.
   * @param {Object} params
   * @param {string} params.policy - Policy template name
   * @param {string} params.claim - Description of the dispute
   * @param {string} params.claimant - Party filing the claim
   * @param {string} params.respondent - Party being claimed against
   * @param {string} [params.desiredRemedy='full_refund'] - What claimant wants
   * @param {Array} [params.evidence=[]] - Evidence items
   * @param {Object} [params.metadata={}] - Structured facts
   * @param {Object} [params.contract={}] - Contract details
   * @returns {Promise<Object>} Ruling object
   *
   * @example
   * const ruling = await court.resolve({
   *   policy: 'sla-monitoring',
   *   claim: 'SLA violation',
   *   claimant: 'client',
   *   respondent: 'provider',
   *   metadata: {
   *     required_uptime: 99.9,
   *     actual_uptime: 98.5,
   *     monitoring_period_confirmed: true
   *   },
   *   evidence: [
   *     { type: 'contract', source: 'sla.pdf',
   *       timestamp: '2026-06-01',
   *       claimed_fact: 'SLA requires 99.9 percent uptime' }
   *   ]
   * });
   */
  async resolve(params) {
    const payload = {
      policy: params.policy,
      claim: params.claim,
      claimant: params.claimant,
      respondent: params.respondent,
      desired_remedy: params.desiredRemedy || 'full_refund',
      evidence: params.evidence || [],
    };
    if (params.metadata) payload.metadata = params.metadata;
    if (params.contract) payload.contract = params.contract;

    const raw = await this._post('/v1/disputes', payload);

    return {
      caseId: raw.case_id || '',
      status: raw.status || '',
      matchedRule: raw.matched_rule_id || '',
      remedy: raw.remedy || '',
      confidence: raw.confidence || '',
      rulingText: raw.ruling || '',
      reasoning: raw.reasoning || '',
      policyName: raw.policy_name || '',
      evidenceScores: raw.evidence_scores || [],
      factsEstablished: raw.facts_established || [],
      factsDisputed: raw.facts_disputed || [],
      factsUnknown: raw.facts_unknown || [],
      ruledAt: raw.ruled_at || '',
      engineVersion: raw.engine_version || '',
      raw,
    };
  }

  /** List all available policy templates. */
  async listPolicies() {
    return await this._get('/v1/policies');
  }

  /** Get details of a specific policy template. */
  async getPolicy(name) {
    return await this._get(`/v1/policies/${name}`);
  }

  /** Look up a previous verdict by case ID. */
  async getVerdict(caseId) {
    return await this._get(`/v1/cases/${caseId}`);
  }

  /** List all stored verdicts. */
  async listVerdicts() {
    return await this._get('/v1/verdicts');
  }

  /** Check API health. */
  async health() {
    return await this._get('/health');
  }
}

/**
 * Convenience function: resolve a dispute in one call.
 * @example
 * import { resolveDispute } from './agentcourt.js';
 * const ruling = await resolveDispute({
 *   policy: 'freelance-delivery',
 *   claim: 'Work never delivered',
 *   claimant: 'buyer',
 *   respondent: 'seller',
 *   evidence: [...]
 * });
 */
export async function resolveDispute(params) {
  return new AgentCourt().resolve(params);
}

// ─── CommonJS fallback (for Node require) ────────────────────────────────
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { AgentCourt, resolveDispute };
}

// ─── Example ──────────────────────────────────────────────────────────────

// Uncomment to test:
/*
async function main() {
  const court = new AgentCourt();
  const health = await court.health();
  console.log(`API: ${health.status} | ${health.policies.length} policies`);

  const ruling = await court.resolve({
    policy: 'physical-commerce',
    claim: 'Wrong product delivered',
    claimant: 'shopping_agent',
    respondent: 'merchant_store',
    desiredRemedy: 'full_refund',
    contract: {
      parties: ['shopping_agent', 'merchant_store'],
      obligations: ['Deliver correct product'],
      deadlines: ['2026-06-25'],
    },
    metadata: {
      delivery_confirmed: true,
      received_matches_order: false,
      ordered_product: 'Blue sneakers size 10',
      received_product: 'Red sneakers size 10',
    },
    evidence: [
      { type: 'contract', source: 'order', timestamp: '2026-06-20',
        claimed_fact: 'Ordered blue sneakers' },
      { type: 'log', source: 'photo', timestamp: '2026-06-22',
        claimed_fact: 'Received red sneakers' },
    ],
  });
  console.log(`Remedy: ${ruling.remedy} (${ruling.confidence})`);
}
main().catch(console.error);
*/
